crates/emitter/src/reference_op_emitter.rs

```diff
@@ -0,0 +1,452 @@
+        //              [stack]
+
+        (self.obj)(emitter)?;
+        //              [stack] OBJ
+
+        (self.key)(emitter)?;
+        //              [stack] OBJ KEY
+
+        Ok(AssignmentReference::new(AssignmentReferenceKind::Elem))
+    }
+}
+
+// Struct for emitting bytecode for call `callee(arguments)` operation.
+pub struct CallEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<CallReference, EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<usize, EmitError>,
+{
+    pub callee: F1,
+    pub arguments: F2,
+}
+impl<F1, F2> CallEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<CallReference, EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<usize, EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        //              [stack]
+
+        let reference = (self.callee)(emitter)?;
+        //              [stack] CALLEE THIS
+
+        // FIXME: Support spread.
+        let len = (self.arguments)(emitter)?;
+        //              [stack] CALLEE THIS ARGS...
+
+        match reference.kind {
+            CallKind::Normal => {
+                emitter.emit.call(len as u16);
+                //      [stack] VAL
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// Struct for emitting bytecode for `new callee(arguments)` operation.
+pub struct NewEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<usize, EmitError>,
+{
+    pub callee: F1,
+    pub arguments: F2,
+}
+impl<F1, F2> NewEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<usize, EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        //              [stack]
+
+        (self.callee)(emitter)?;
+        //              [stack] CALLEE
+
+        emitter.emit.is_constructing();
+        //              [stack] CALLEE JS_IS_CONSTRUCTING
+
+        // FIXME: Support spread.
+        let len = (self.arguments)(emitter)?;
+        //              [stack] CALLEE JS_IS_CONSTRUCTING ARGS...
+
+        emitter.emit.dup_at(len as u32 + 1);
+        //              [stack] CALLEE JS_IS_CONSTRUCTING ARGS... CALLEE
+
+        emitter.emit.new_(len as u16);
+        //              [stack] VAL
+
+        Ok(())
+    }
+}
+
+// Struct for emitting bytecode for assignment `lhs = rhs` operation.
+pub struct AssignmentEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<AssignmentReference, EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub lhs: F1,
+    pub rhs: F2,
+}
+impl<F1, F2> AssignmentEmitter<F1, F2>
+where
+    F1: Fn(&mut AstEmitter) -> Result<AssignmentReference, EmitError>,
+    F2: Fn(&mut AstEmitter) -> Result<(), EmitError>,
+{
+    pub fn emit(self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
+        //              [stack]
+
+        let reference = (self.lhs)(emitter)?;
```

Comment:
This design makes a lot of sense.

(Debug assertions about the stack are I think still worth doing here, although you'll have to add a method:)

```
impl AssignmentReference {
    #[cfg(debug)]
    fn stack_slots(&self) -> usize {
        match self {
            AssignmentReferenceKind::GlobalName(..) => 1,
            AssignmentReferenceKind::Prop(..) => 1,
            AssignmentReferenceKind::Elem => 2,
        }
    } 
}
```